### PR TITLE
fix Issue 20691 - Converting scope static array to scope dynamic arra…

### PR DESCRIPTION
**Pull request by WalterBright**
…y should be error

### CODE REVIEW

**Comment by Geod24** [comment]
Added a link in 20691 to [20505](https://issues.dlang.org/show_bug.cgi?id=20505) (why the duplication ??).
So do I understand correctly that the reasoning is that there can only be one level of scope-ness ?

**Comment by WalterBright** [comment]
> why the duplication ??

I originally thought it was a different problem.

> the reasoning is that there can only be one level of scope-ness ?

Yes. Scope is not transitive.

**Comment by WalterBright** [comment]
Unfortunately, several projects in buildkite are relying on this compiler bug.

**Comment by Geod24** [comment]
> Yes. Scope is not transitive.

Then why do `struct` allow one to store a `scope` slice in them ?

Also, can you renamed the commit title to `Fix issue 20505 - [...]`, since 20691 was a duplicate ?

**Comment by Geod24** [comment]
> Unfortunately, several projects in buildkite are relying on this compiler bug.

We discussed this a million time, but I'm going to repeat it anyway: *anyone* trying to do something useful with `scope` (that is, past identity functions) expect it to be transitive.
If `scope` is not transitive, then how do we deal with struct storing slices ?

I tested this PR with some sample code, and this still compiles:
```D
import std.stdio;

struct Bar
{
    const(char)[] v;
}

struct Store
{
    Bar[] v;
}

void main () @safe
{
    auto v = func1();
    writeln(v);
}

const(char)[] func1 () @safe
{
    char[10] str = 'a';
    return func2(str);
}

const(char)[] func2 (scope const(char)[] b) @safe
{
    Bar[10] array = Bar(b);
    return func3(array, b);
}

const(char)[] func3 (scope Bar[] array, scope const(char)[] b) @safe
{
    Store st = Store(array);
    return st.v[0].v;
}
```

**Comment by Geod24** [comment]
Any update on this ?

**Comment by WalterBright** [comment]
> @Geod24 this still compiles

please add bug report.

**Comment by Geod24** [comment]
> please add bug report.

I did, it's the one you are trying to fix (https://issues.dlang.org/show_bug.cgi?id=20505).

**Comment by WalterBright** [comment]
@Geod24 as for transitive scope, while a good idea, that would be a fairly major project, way beyond the scope of this PR. It is likely to also break much existing code.

**Comment by Geod24** [comment]
@WalterBright :

1) I agree with you, having transitive `scope` would be a fairly major undertaking, and TBH I'm not sure it would allow us to have `@safe` RC either;

2) By definition, code that is triggered by a `-preview` is experimental. So at the moment, whatever change we decide to go with, as long as it only triggers when `-preview=dip1000` is used, it's not a breaking change;

I'll merge the druntime PR in the meantime.

**Comment by dkorpel** [comment]
Now that the blocker from Phobos has been fixed, can this be merged? I assume the failing test runs are old and should go away after a rebase.

**Comment by dkorpel** [comment]
@WalterBright @RazvanN7 Can you rerun the tests?

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -12490,13 +12490,15 @@ bool checkAddressVar(Scope* sc, Expression exp, VarDeclaration v)
                 v.storage_class &= ~STC.maybescope;
                 v.doNotInferScope = true;
                 if (exp.type.hasPointers() && v.storage_class & STC.scope_ &&
+                    !(v.storage_class & STC.temp) &&
```
Comment (dkorpel) [inline src/dmd/expressionsem.d:12493]: What's this for?

Comment (MoonlightSentinel) [reply src/dmd/expressionsem.d:12493]: Maybe a remnant of the old bug that caused `enclosesLifetimeOf` to return wrong results for `VarDeclaration`s created during semantic?

**Comment by dkorpel** [comment]
Ping @WalterBright @RazvanN7

**Comment by WalterBright** [comment]
Looks like the problem is in phobos:
```
std/process.d(1960): Error: cannot take address of `scope` local `args` in `@safe` function `spawnShell`
std/process.d(1960): Error: cannot cast expression `args` of type `const(char)[][3]` to `const(char[])[]`
```

**Comment by dkorpel** [comment]
> Looks like the problem is in phobos:

So the error is still there. I thought it would be fixed by https://github.com/dlang/phobos/pull/8119

**Review by RazvanN7 — APPROVED** [review]

**Comment by thewilsonator** [comment]
Removing auto merge this has failures all across build kite

**Pull request merged**